Spigot-Server-Patches/0752-add-get-set-biome-for-namespaced-keys.patch

```diff
@@ -0,0 +1,122 @@
+             return new IllegalStateException("Missing registry: " + resourcekey);
+@@ -119,7 +120,7 @@ public abstract class IRegistryCustom {
+     }
+ 
+     private static <R extends IRegistry<?>> void a(IRegistryCustom.Dimension iregistrycustom_dimension, ResourceKey<R> resourcekey) {
+-        IRegistry<R> iregistry = RegistryGeneration.b;
++        IRegistry<R> iregistry = (IRegistry<R>) RegistryGeneration.b; // paper - decompile fix
+         IRegistry<?> iregistry1 = (IRegistry) iregistry.a(resourcekey);
+ 
+         if (iregistry1 == null) {
+@@ -213,14 +214,12 @@ public abstract class IRegistryCustom {
+         }
+ 
+         private static <E> RegistryMaterials<?> d(ResourceKey<? extends IRegistry<?>> resourcekey) {
+-            return new RegistryMaterials<>(resourcekey, Lifecycle.stable());
++            return new RegistryMaterials(resourcekey, Lifecycle.stable()); // Paper - decompile fix
+         }
+ 
+         @Override
+         public <E> Optional<IRegistryWritable<E>> a(ResourceKey<? extends IRegistry<E>> resourcekey) {
+-            return Optional.ofNullable(this.b.get(resourcekey)).map((registrymaterials) -> {
+-                return registrymaterials;
+-            });
++            return Optional.ofNullable((IRegistryWritable<E>) this.b.get(resourcekey)); // Paper - decompile fix
+         }
+     }
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+index 9d858f91828d6c2787ff1dc677a247bcab172701..be4b92140a40703c9695d6c69bddfee34a91e066 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -1099,6 +1099,28 @@ public class CraftWorld implements World {
+         return CraftBlock.biomeBaseToBiome(getHandle().r().b(IRegistry.ay), this.world.getBiome(x >> 2, y >> 2, z >> 2));
+     }
+ 
++    // Paper start
++    @Override
++    public org.bukkit.NamespacedKey getBiomeKey(int x, int y, int z) {
++        return CraftBlock.biomeBaseToKey(getHandle().getWorldCustomRegistry().getRegistry(IRegistry.getBiomeKeys()), this.world.getBiome(x >> 2, y >> 2, z >> 2));
++    }
++
++    @Override
++    public void setBiome(int x, int y, int z, org.bukkit.NamespacedKey key) {
++        BiomeBase base = CraftBlock.biomeToBiomeBase(getHandle().getWorldCustomRegistry().getRegistry(IRegistry.getBiomeKeys()), key);
++        if (base == null) {
++            throw new NullPointerException("Invalid biome key, biome was null.");
++        }
++        BlockPosition pos = new BlockPosition(x, 0, z);
++        if (this.world.isLoaded(pos)) {
++            net.minecraft.world.level.chunk.Chunk chunk = this.world.getChunkAtWorldCoords(pos);
++            if (chunk != null) {
++                chunk.getBiomeIndex().setBiome(x >> 2, y >> 2, z >> 2, base);
++                chunk.markDirty(); // SPIGOT-2890
++            }
++        }
++    }
```

Comment:
Instead of coping the logic, make the existing setBiome method just call this method passing in its namespaced key.